flytekit/image_spec/image_spec.py

```diff
@@ -329,6 +330,37 @@ def force_push(self) -> "ImageSpec":
 
         return copied_image_spec
 
+    @classmethod
+    def from_env(cls, *, pinned_packages: Optional[List[str]] = None, **kwargs) -> "ImageSpec":
+        """Create ImageSpec with the environment's Python version and packages pinned to the ones in the environment."""
+
+        from importlib.metadata import version
+
+        # Invalid kwargs when using `ImageSpec.from_env`
+        invalid_kwargs = ["python_version"]
+        for invalid_kwarg in invalid_kwargs:
+            if invalid_kwarg in kwargs and kwargs[invalid_kwarg] is not None:
+                msg = (
+                    f"{invalid_kwarg} can not be used with `from_env` because it will be inferred from the environment"
+                )
+                raise ValueError(msg)
+
+        version_info = sys.version_info
+        python_version = f"{version_info.major}.{version_info.minor}"
+
+        if "packages" in kwargs:
+            packages = kwargs.pop("packages")
+        else:
+            packages = []
+
+        pinned_packages = pinned_packages or []
+
+        for package_to_pin in pinned_packages:
+            package_version = version(package_to_pin)
+            packages.append(f"{package_to_pin}=={package_version}")
```

Comment:
should there be any special handling for when there are overlapping packages in `pinned_packages` and `packages`?

e.g. replace the unpinned package with the pinned package and logging via `logger.info`?

Reply:
There is no issue with overlapping unless there is a conflict. For example, this installs `flytekit==1.13.8`:

```bash
uv pip install flytekit flytekit==1.13.8
```

If there is a conflict, then the builder will error:

```bash
uv pip install flytekit==1.13.11 flytekit==1.13.12
  × No solution found when resolving dependencies:
  ╰─▶ Because you require flytekit==1.13.11 and flytekit==1.13.12, we can conclude that your requirements are
      unsatisfiable.
```

Given that we also have `requirements.txt`, there can be conflicts there too. I rather have the not do the error checking here.

As an alternative, I considered disallowing `packages` & `requirements` here, but it seemed to restrictive.